openwisp_monitoring/monitoring/charts.py

```diff
@@ -131,6 +131,60 @@
             )
         },
     },
+    'memory': {
+        'type': 'scatter',
+        'title': _('Memory Usage'),
+        'description': _('Percentage of memory (RAM) being used.'),
+        'summary_labels': [_('Memory Usage')],
+        'unit': '%',
+        'colors': [DEFAULT_COLORS[4]],
+        'order': 250,
+        'query': {
+            'influxdb': (
+                "SELECT 100 * (1 - ((MEDIAN(free_memory) + MEDIAN(buffered_memory)) "
+                "/ MEDIAN(total_memory))) AS memory_usage "
```

Comment:
@nepython @PabloCastellano if there's a way to simplify this query, I would do it as well, with the goal of removing load on influxdb when retrieving the value from the DB.

Reply:
It can be done by doing storing the percentage itself in influxdb and retrieving it like I had been doing previously. Also, in that case, only one value will get stored in influxdb. @nemesisdesign @PabloCastellano, should I do it that way then?

Reply:
What about returning `available_memory` instead?

Reply:
That might be better as the data stored might be usable somewhere else too later on (percentage might not be usable) but this reminds me did we resolve the threshold default value? I can't recall :sweat_smile:. The threshold would compare with the field `available_memory` which will vary from device to device so we can use a default threshold value only with percentage I think. The other work around might be setting some percentage of total memory as the threshold value to compare with.

Reply:
correct, for the problem of the threshold, we need to keep it in percentage.
Let's keep the raw data, but also go back to store the processed data, which will be faster.

Further optimizations can be done only after trying it in the real world, but storage issues can be dealt with more easily than RAM/CPU usage to do the calculations.

Reply:
Last time I checked thresholds worked fine with percentages (0-1).

We are already storing the percentage as the *primary* value:

```
            metric.write(used_bytes / size_bytes)
```

Reply:
@PabloCastellano , that's for `Disk` and it's correct. We were talking about `Memory` though. Also, I had one change to recommend, shall we store,
```python
threshold = cpus*DEVICE_RESOURCES_THRESHOLDS['load']
```
Since, currently it's correct threshold only for single core devices. This might be needed as we are **not** storing metric value for load as `load_0 / cpus`

Reply:
You are right but same applies for memory:

```
used_memory = 1 - (memory['free'] + memory['buffered']) / memory['total']
metric.write(used_memory, extra_values=extra_values)
```

used_memory is a float between 0 and 1

> threshold = cpus*DEVICE_RESOURCES_THRESHOLDS['load']

Instead of doing this, I would save the percentage in the *primary* key of the Metric. It could be named `cpu_usage` and save `load_1` among the rest of extra values. IMO this way it is clearer because you understand that the threshold is 90% of the usage (0.9). Seeing instead a threshold of 1,8 because it's a 2-cores device is confusing.